### ISSUE TITLE
Regression with expression templates and bitfields.

### ISSUE THREAD

**Issue by rkaminsk** [catchorg/Catch2#1027]
## Description
The program below fails to compile (but did with version 1.9.6). The first `REQUIRE` compiles just fine, while the second does not:
```c++
#include "catch.hpp"
void test() {
    struct Y { 
        uint32_t v : 1;
    };  
    Y y{0};
    REQUIRE(y.v == 0);
    REQUIRE(0 == y.v);
} 
```

The problem is the following operator (passing bitfields by reference is not possible):
```c++
template<typename LhsT>
template<typename RhsT>
auto ExprLhs<LhsT>::operator == ( RhsT&& rhs ) -> BinaryExpr<LhsT, RhsT&> const {
    return BinaryExpr<LhsT, RhsT&>( compareEqual( m_lhs, rhs ), m_lhs, "==", rhs ); 
}
```

Replacing it with the following version fixes the issue:
```c++
template<typename LhsT>
template<typename RhsT>
auto ExprLhs<LhsT>::operator == ( RhsT const & rhs ) -> BinaryExpr<LhsT, RhsT const &> const {
    return BinaryExpr<LhsT, RhsT const &>( compareEqual( m_lhs, rhs ), m_lhs, "==", rhs );
}
````

I have not looked deeper into the code. Is there a reason to not explicitely use a const reference here?

### Extra information
* Catch version: v2.0.0-develop.4 and v1.10
* Compiler+version: clang/gcc

**Comment by lightmare** [comment]
Couldn't reproduce with v1.10.0, only with catch2 branch.

I haven't been watching Catch for a loooong time, but catch2 does not seem fully rvalue-reference aware, yet. It's not capturing rvalue references where it could, and it's not properly forwarding universal references where it should (actually, `std::forward` doesn't appear anywhere in the code).

Here's a fix I proposed long ago, capturing not just `bool` but all fundamental types by value:

```diff
diff --git a/include/internal/catch_decomposer.h b/include/internal/catch_decomposer.h
index 786a052..7090b60 100644
--- a/include/internal/catch_decomposer.h
+++ b/include/internal/catch_decomposer.h
@@ -53,9 +53,9 @@ namespace Catch {
     public:
         BinaryExpr( bool comparisonResult, LhsT lhs, StringRef op, RhsT rhs )
         :   m_result( comparisonResult ),
-            m_lhs( lhs ),
+            m_lhs( std::forward<LhsT>(lhs) ),
             m_op( op ),
-            m_rhs( rhs )
+            m_rhs( std::forward<RhsT>(rhs) )
         {}
     };
 
@@ -98,6 +98,8 @@ namespace Catch {
     template<typename T>
     auto compareNotEqual( long lhs, T* const& rhs ) -> bool { return reinterpret_cast<void const*>( lhs ) != rhs; }
 
+    template <typename T>
+    using capture_by_value = std::is_fundamental<typename std::remove_cv<typename std::remove_reference<T>::type>::type>;
 
     template<typename LhsT>
     class ExprLhs {
@@ -105,12 +107,16 @@ namespace Catch {
     public:
         ExprLhs( LhsT lhs ) : m_lhs( lhs ) {}
 
-        template<typename RhsT>
-        auto operator == ( RhsT&& rhs ) -> BinaryExpr<LhsT, RhsT&> const {
-            return BinaryExpr<LhsT, RhsT&>( compareEqual( m_lhs, rhs ), m_lhs, "==", rhs );
+        template<typename RhsT, typename = typename std::enable_if<!capture_by_value<RhsT>::value>::type>
+        auto operator == ( RhsT&& rhs ) -> BinaryExpr<LhsT, RhsT&&> const {
+            return BinaryExpr<LhsT, RhsT&&>( compareEqual( m_lhs, rhs ),
+                                             std::forward<LhsT>( m_lhs ), "==",
+                                             std::forward<RhsT>( rhs ) );
         }
-        auto operator == ( bool rhs ) -> BinaryExpr<LhsT, bool> const {
-            return BinaryExpr<LhsT, bool>( m_lhs == rhs, m_lhs, "==", rhs );
+        template<typename RhsT, typename = typename std::enable_if<capture_by_value<RhsT>::value>::type>
+        auto operator == ( RhsT rhs ) -> BinaryExpr<LhsT, RhsT> const {
+            return BinaryExpr<LhsT, RhsT>( compareEqual( m_lhs, rhs ),
+                                           std::forward<LhsT>( m_lhs ), "==", rhs );
         }
 
         template<typename RhsT>
```

... and similarly for other operators.

Note: at first I wanted to use `is_scalar` in place of `is_fundamental`, but then the operator becomes ambiguous because both overloads match when `rhs` is a string literal due to implicit conversions to `const char*` and `std::string`.

**Comment by rkaminsk** [comment]
Some comments:
- Yes, with catch 1 this works. Sorry for wrongly reporting it doesn't. Catch 1 uses const references, which makes sense with comparison operators.
- I don't think that the patch uses forwarding/universal references properly. Forwarding (potentially moving) something and using it more than once in an expression cannot be right.
- The catch 2 code does not look right to me either. Is the intention to also support non-const references here? Then just using `RhsT&` instead of `RhsT&&` should do the trick (which unfortunately still requires special treatment to handle bitfields). I am just guessing what is intended.

I found this article about the topic nice: http://eli.thegreenplace.net/2014/perfect-forwarding-and-universal-references-in-c/

**Comment by lightmare** [comment]
You're right. I assumed that the operator taking universal reference was intentional, and tried to stick proper reference type into the `BinaryExpr`, without thinking it through.

Now I agree that taking `RhsT const&` should be enough. Unless someone wants to test custom `operator==` overloads working differently for const/non-const lvalue/rvalue objects (I can't imagine how that'd be useful), in which case they'd need Catch operators to capture the type precisely.

I'll update the diff above to not clutter the issue too much. Had to fix one test that had non-cost conversion operator.

**Comment by rkaminsk** [comment]
Passing a bitfield via const reference is no problem. Now, we do not even need the enable_if anymore. ;)

**Comment by horenmar** [comment]
I spent some time looking over this and I am still not sure why it is using mutable references.

We used to support non-const comparison operators, but to support those we const-casted away the constness anyway. I am going to throw it over at @philsquared as the Catch 2 rewritten decomposer is his work, before making changes.

**Comment by philsquared** [comment]
I've just pushed a change so that it takes the RHS by const reference.

When I was rewriting it (from the original Catch 1.x version) I was trying to be const/ non-const agnostic (to avoid having to `const_cast` at the bottom). My reasoning was that it's easy to write a comparison operator overload and forget to make it const. The person writing the test may not even have any control over that. And it would be a difference in behaviour from doing the comparison without the expression decomposition - which I was trying to avoid.
Anyway, I ended up having to make the LHS a const ref anyway, and this issue has shown that my attempt to hold out on the RHS was misguided. So I'm letting go on this one.

The universal references where never about perfect forwarding - no values are being stored as we're only ever consulting the values while they are still on the stack.
